docker/stan-wasm-server/logic/locking.py

```diff
@@ -1,37 +1,36 @@
-from os import getpid
 import time
 from pathlib import Path
-
-def get_nonce():
-    return str(getpid()) + str(time.time())
+from contextlib import contextmanager
 
 
 def _get_compilation_lockfile_name(model_dir: Path):
     p = model_dir / "running.txt"
     return p
 
 
-def acquire_compilation_lock(model_dir: Path, nonce: str):
+def wait_until_free(model_dir: Path):
     lockfile = _get_compilation_lockfile_name(model_dir)
+    while lockfile.is_file():
+        time.sleep(0.5)
+
+
+@contextmanager
+def compilation_output_lock(model_dir: Path):
+    lockfile = _get_compilation_lockfile_name(model_dir)
+
+    acquired = False
     try:
         with lockfile.open(mode='x') as file:
-            file.write(nonce)
-        contents = lockfile.read_text()
-        return contents == nonce
-    except:
-        return False
+            file.write("locked")
 
+        acquired = True
+    except:
+        acquired = False
 
-def release_compilation_lock(model_dir: Path, nonce: str):
-    lockfile = _get_compilation_lockfile_name(model_dir)
     try:
-        contents = lockfile.read_text()
-        if contents == nonce:
+        yield acquired
+    finally:
+        if acquired:
```

Comment:
Maybe feels better to do this as one try-catch instead of two? Also the `finally` is conditioned on successfully completing the `try` block. Maybe:

```python
lockfile = _get_compilation_lockfile_name(model_dir)

try:
    with lockfile.open(mode='x') as file:
        file.write("locked")
        yield True
        lockfile.unlink(missing_ok=True)
    except:
        yield False
```

Hm, what happens if there's an exception elsewhere after we `yield`--would that result in failure to unlock? If so we'd need to put the `finally` back.

Reply:
We need two `try`s somewhere - they can be nested or in sequence, but the unlock needs to happen in a `finally` in case anything throws during the `yield`

Reply:
Made this into a nested try, which looks cleaner IMO